Add @workspace doc (#7109)
microsoft/vscode-docs: docs/copilot/workspace-context.md

---
Order: 7
Area: copilot
TOCTitle: Workspace Context
ContentId: c77dcce9-4ba9-40ac-8ae5-2df855088090
PageTitle: Chat using @workspace Context References
DateApproved: 02/28/2024
MetaDescription: Overview of how the @workspace Copilot chat participants works and how it manages context.
---
# How the @workspace Copilot chat participant works

> 

Reviewer: Should we first have a sentence or two intro on what participants are?

Reviewer 2: Afaik we should not open the H1 with chat participant, and optimize for SEO and a more natural language

- Finding existing code in your codebase:
  - `"@workspace where is database connecting string configured?"`
  - `"@workspace is there a library for date validation?"`
  - `"Where are tests defined?"`
- Making plans for larger codebase edits:
  - `"@workspace how can I add a hover text to a button?"`
  - `"@workspace add date validation to #selection"`
  - `"@workspace how do I add a new a new API route?"`
- Explaining higher-level concepts in a codebase:
  - `"@workspace how is authentication implemented?"`
  - `"@workspace which API routes does this service provide?"`
  - `"How do I build this project?"`

## What context does @workspace have?

The goal of the `@workspace` is to provide answers to questions about your full workspace that you have open in VS Code. It indexes different sources of information and applies different search strategies to come up with the most relevant context:

- All files in the VS Code workspace, except for files that are included in `.gitignore` files
- Workspace structure, such as folder and file names
- GitHub's code search index, if the workspace is a GitHub repository and indexed by code search
- Symbols and definitions in the workspace

> - Currently selected text (`#selection`) or visible text in the active editor (`#editor`)

Reviewer: Do we have docs on `#variables`? It might be useful to link to some docs on them and/or explain a bit more about them here in case folks haven't seen this syntax.

Author: Yes, these are explained in the updated Copilot docs (#7090 )

**Note**: `.gitignore` is overridden if you have a file open or have text selected in an ignored file.

> ## How does @workspace work?

Reviewer: This heading feels like it duplicates the first heading of the doc `# How the @workspace Copilot chat participant works`. Should the first heading be changed to something more general like `# @workspace Copilot chat participant`?

Reviewer: After reading this section, I wonder if it'd also make more sense to specifically mention context in the header, i.e.:

```suggestion
## How does @workspace use this context?
```

I mention this since I thought this would be distinct from the above section describing context, but instead it seems to describe how that context we just mentioned is actually used.

First, `@workspace` uses GitHub Copilot to disambiguate the user's prompt and to get input for the different search strategies that will produce the right amount of context.

Next, it performs a combination of search strategies, such as [GitHub's code search index](https://github.blog/2023-02-06-the-technology-behind-githubs-new-code-search), a lexical text searches over the local index to find local, uncommitted changes, and VS Code's language intelligence to add details like function signatures, parameters, and more.

Finally, all of these pieces of context are ranked, sliced, and summarized by `@workspace`, and then sent off to GitHub Copilot to answer the user's request.

After the Copilot response is received, `@workspace` processes the response and adds references to files, file ranges, and symbols. This enables users to link directly from the chat response to the corresponding information in their codebase.

## Tips for using @workspace

> - Be specific in your questions and use terms and concepts that are potentially present in your codebase

Reviewer: Could be great to give an example of a specific vs non-specific question.

- Use chat variables to expand the explicit context on top of what is automatically included from the codebase, such as `#editor`, `#selection`, or `#file`
- Don't expect statistical analysis, such as "how many times is this function called?" or "how many times is this variable used?"

> ## Context for @workspace slash commands

Reviewer: I would've expected this section to come more directly after the other context section(s). I thought that since the section prior was describing tips, we were already basically done with the doc (since I expect docs to conclude with tips and tricks, FAQ, etc).

> 

Reviewer: Could be helpful to introduce a few things:

- What are slash commands
- The fact that certain slash commands are contributed by / used with `@workspace`

If you're using slash commands, then they define their own context that is optimized for their specific task:

| Command    | Context |
| ---------- | ------- |

> | `/explain` | <ul><li>Starts with the text selection in the active editor (#selection). To optimize the Copilot chat responses, make sure to expand the text selection to include any relevant information to help Copilot provide a useful response.</li></ul><ul><li>Looks up the implementations of referenced symbols such as functions and classes, leading to more accurate and useful explanations.</li></ul> |

Reviewer: ```suggestion
| `/explain` | <ul><li>Starts with the text selection in the active editor (`#selection`). To optimize the Copilot chat responses, make sure to expand the text selection to include any relevant information to help Copilot provide a useful response.</li></ul><ul><li>Looks up the implementations of referenced symbols such as functions and classes, leading to more accurate and useful explanations.</li></ul> |
```

| `/test`    | <ul><li>Current text selection in the active editor. If no text is selected, use the contents of the currently active file.</li></ul><ul><li>Related existing test files, to understand existing tests and best practices.</li></ul> |
| `/fix`     | <ul><li>Current text selection in the active editor. If no text is selected, use the currently visible text in the editor.</li></ul><ul><li>Errors and referenced symbols to understand what needs to be fixed and how.</li></ul> |

> You can explicitly expand the context by using chat variables, such as `#editor`, `#selection`, or `#file` in your chat prompt. For example, to get an explanation of the currently visible code in the editor, use this chat prompt: `@workspace /explain #editor`.

Reviewer: I think this explanation of `#variables` is what I was looking for earlier in the doc!

---
Order: 7
Area: copilot
TOCTitle: Workspace Context
ContentId: c77dcce9-4ba9-40ac-8ae5-2df855088090
PageTitle: Chat using @workspace Context References
DateApproved: 02/28/2024
MetaDescription: How to use Copilot's @workspace chat to ask questions against your entire codebase.
---
# Making Copilot Chat an expert in your workspace

> Referencing `@workspace` in Copilot Chat lets you ask questions about your entire codebase. Based on the question Copilot intelligently retrieves relevant files and symbols, which it then references in its answer as links and code examples. Grounded in `@workspace` references, Copilot Chat becomes a domain expert for tasks like:

Reviewer: nit to potentially help with flow

```suggestion
Referencing `@workspace` in Copilot Chat lets you ask questions about your entire codebase. Based on the question, Copilot intelligently retrieves relevant files and symbols, which it then references in its answer as links and code examples. Grounded in `@workspace` references, Copilot Chat becomes a domain expert for tasks like:
```

- Finding existing code in your codebase:
  - `"@workspace where is database connecting string configured?"` - Explains where and how the database connection is configured
  - `"@workspace how can I validate a date?"` - Finds existing date validation helpers in the codebase
  - `"@workspace where are tests defined?"` - Provides the location of test suites, cases, and related references and configurations
- Making plans for complex code edits:
  - `"@workspace how can I add a rich tooltip to a button?"` - Provides a plan for using the existing tooltip component with button elements
  - `"@workspace add date validation to #selection"` - Plans how to apply the existing date validation to the selected code
  - `"@workspace add a new API route for the forgot password form"` - Outlines where to add the new route and how to connect it to the existing code
- Explaining higher-level concepts in a codebase:
  - `"@workspace how is authentication implemented?"` - Overview of the authentication flow and references to the relevant code
  - `"@workspace which API routes depend on this service?"` - Lists the routes that use the service in the selected code

>   - `"How do I build this #codebase?"` - List the steps to build the project based on documentation, scripts, and configurations

Reviewer: Sorry if I missed a reply to my previous comment here - this example stands out to me because it doesn't specifically use `@workspace`. Might want to explain that using `#codebase` is an alternative?

## What sources does `@workspace` use for context?

To answer your question, `@workspace` searches through the same sources a developer would use when navigating a codebase in VS Code:

> - All files in the workspace, except for files that are included in `.gitignore` files

Reviewer 3: ```suggestion
- All files in the workspace, except for files that are ignored by a `.gitignore` file
```

- Symbols and definitions in the workspace
- Currently selected text or visible text in the active editor

**Note**: `.gitignore` is bypassed if you have a file open or have text selected within an ignored file.

## How does `@workspace` find the most relevant context

Your full VS Code workspace can be too large to pass entirely to GitHub Copilot for responding to your chat prompt. Instead, `@workspace` extracts the most relevant information from the different context sources to ground Copilot's answer.

First, `@workspace` determines which information is needed to answer your question, also including the conversation history, workspace structure, and currently selected code.

Next, it performs a combination of search strategies, such as [GitHub's code search index](https://github.blog/2023-02-06-the-technology-behind-githubs-new-code-search), a lexical text searches over the local index to find local, uncommitted changes, and VS Code's language intelligence to add details like function signatures, parameters, and more.

> Finally, all of these pieces of context are ranked, sliced, and summarized by `@workspace`, and then used by GitHub Copilot to answer your question. The responses is marked up with references to files, file ranges, and symbols. This enables you to link directly from the chat response to the corresponding information in your codebase.

Reviewer 3: I think we should keep this higher level. The first part is good, but for steps 2 and 3 I'd avoid talking about specifics like lexical searching or slicing, which may be confusing and also aren't really technically correct

Maybe something more like (needs word cleanup):

- First `@workspace` determines what context (file names, symbol names, code excepts, ...) from the workspace may help answer the question. 

- Next it collects the context using a variety of different approaches. This includes finding relevant snippets of code by searching locally or using Github Code search. We also collect context using VS Code's language IntelliSense's

- Finally we pass this context to copilot. If the context is large, we only pass the most potentially relevant parts of the context. The code snippets that were passed to copilot are listed as references in the response

| `/test`    | <ul><li>Current text selection in the active editor. If no text is selected, use the contents of the currently active file.</li></ul><ul><li>Related existing test files, to understand existing tests and best practices.</li></ul> |
| `/fix`     | <ul><li>Current text selection in the active editor. If no text is selected, use the currently visible text in the editor.</li></ul><ul><li>Errors and referenced symbols to understand what needs to be fixed and how.</li></ul> |

You can explicitly expand the context by using chat variables, such as `#editor`, `#selection`, or `#file` in your chat prompt. For example, to fix an error in the current file based on a pattern from another file, use this chat prompt: `@workspace /fix linting error in the style of #file:form.ts`.

## Tips for using `@workspace`

The way you phrase your question can significantly influence the quality of the references `@workspace` provides and the accuracy of the response. To optimize results, consider the following tips:

> - Be specific and detailed in your question, avoiding vague or ambiguous terms like "what does this do" (could be the last answer, current file, or whole project, etc.).

Reviewer: ```suggestion
- Be specific and detailed in your question, avoiding vague or ambiguous terms like "what does this do" (where "this" could be interpreted as the last answer, current file, or whole project, etc.).
```
